File: core/src/nl/t64/game/rpg/screens/inventory/itemslot/ItemSlot.java
package nl.t64.game.rpg.screens.inventory.itemslot;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Stack;
import com.badlogic.gdx.utils.GdxRuntimeException;
import com.badlogic.gdx.utils.Scaling;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.party.InventoryGroup;
import nl.t64.game.rpg.screens.inventory.tooltip.ItemSlotTooltip;

import java.util.Optional;


public abstract class ItemSlot extends Stack {

    private static final String SPRITE_BACKGROUND = "sprites/inventoryslot.png";
    private static final String SPRITE_SELECTED = "sprites/selected.png";
    private static final String NOTHING = "nothing";
    private static final String BASIC = "basic";
    private static final String FINE = "fine";
    private static final String SPECIALIST = "specialist";
    private static final String MASTERWORK = "masterwork";
    private static final String EPIC = "epic";
    private static final String LEGENDARY = "legendary";

    public final int index;
    public final InventoryGroup filterGroup;
    public final ItemSlotTooltip tooltip;
    public final Stack imagesBackground;

    public ItemSlot(int index, InventoryGroup filterGroup, ItemSlotTooltip tooltip) {
        this.index = index;
        this.filterGroup = filterGroup;
        this.tooltip = tooltip;
        this.imagesBackground = createImageBackground();
        addToStack(this.imagesBackground);
    }

    private Stack createImageBackground() {
        var stack = new Stack();
        var texture = Utils.getResourceManager().getTextureAsset(SPRITE_BACKGROUND);
        var background = new Image(texture);
        background.setName("background");
        stack.add(background);
        stack.add(createTierColor(NOTHING));
        return stack;
    }

    public void select() {
        var texture = Utils.getResourceManager().getTextureAsset(SPRITE_SELECTED);
        var selector = new Image(texture);
        selector.setName("selector");
        super.add(selector);
        tooltip.refresh(this);
    }

    public void deselect() {
        if (isSelected()) {
            super.getChildren().pop();
        } else {
            throw new GdxRuntimeException("Tried to deselect an unselected ItemSlot.");
        }
    }

    public boolean isSelected() {
        String name = super.getChildren().peek().getName();
        return name != null && name.equals("selector");
    }

    public Optional<InventoryImage> getPossibleInventoryImage() {
        if (hasItem()) {
            return Optional.of(getCertainInventoryImage());
        }
        return Optional.empty();
    }

    public InventoryImage getCertainInventoryImage() {
        return (InventoryImage) super.getChild(1);
    }

    public abstract void addToStack(Actor actor);

    public abstract boolean hasItem();

    public abstract boolean isOnHero();

    public abstract boolean doesAcceptItem(InventoryImage draggedItem);

    public abstract void clearStack();

    public abstract void putInSlot(InventoryImage draggedItem);

    public abstract int getAmount();

    public abstract void incrementAmountBy(int amount);

    public abstract void decrementAmountBy(int amount);

    void putItemBack(InventoryImage draggedItem) {
        putInSlot(draggedItem);
    }

    int getHalfOfAmount() {
        return (int) Math.floor(getAmount() / 2f);
    }

    public void setItemColor() {
        imagesBackground.getChildren().pop();
        if (hasItem()) {
            imagesBackground.add(createTierColor());
        } else {
            imagesBackground.add(createTierColor(NOTHING));
        }
    }

    private Image createTierColor() {
        String itemId = getCertainInventoryImage().inventoryItem.getId();
        if (itemId.contains(BASIC)) {
            return createTierColor(BASIC);
        } else if (itemId.contains(FINE)) {
            return createTierColor(FINE);
        } else if (itemId.contains(SPECIALIST)) {
            return createTierColor(SPECIALIST);
        } else if (itemId.contains(MASTERWORK)) {
            return createTierColor(MASTERWORK);
        } else if (itemId.contains(EPIC)) {
            return createTierColor(EPIC);
        } else if (itemId.contains(LEGENDARY)) {
            return createTierColor(LEGENDARY);
        } else {
            return createTierColor(NOTHING);
        }
    }

    private Image createTierColor(String tierColor) {
        var textureRegion = Utils.getResourceManager().getAtlasTexture(tierColor);
        var image = new Image(textureRegion);
        image.setScaling(Scaling.none);
        image.setPosition(1f, 1f);
        image.setName("tier color");
        return image;
    }

}

File: core/src/nl/t64/game/rpg/screens/inventory/itemslot/ItemSlot.kt
package nl.t64.game.rpg.screens.inventory.itemslot

import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Stack
import com.badlogic.gdx.utils.GdxRuntimeException
import com.badlogic.gdx.utils.Scaling
import nl.t64.game.rpg.Utils.resourceManager
import nl.t64.game.rpg.components.party.inventory.InventoryGroup
import nl.t64.game.rpg.screens.inventory.tooltip.ItemSlotTooltip
import kotlin.math.floor


private const val SPRITE_BACKGROUND = "sprites/inventoryslot.png"
private const val SPRITE_SELECTED = "sprites/selected.png"
private const val NOTHING = "nothing"
private const val BASIC = "basic"
private const val FINE = "fine"
private const val SPECIALIST = "specialist"
private const val MASTERWORK = "masterwork"
private const val EPIC = "epic"
private const val LEGENDARY = "legendary"

abstract class ItemSlot(
    val index: Int,
    val filterGroup: InventoryGroup,
    val tooltip: ItemSlotTooltip
) : Stack() {

    val imagesBackground: Stack = createImageBackground().also {
        addToStack(it)
    }

    private fun createImageBackground(): Stack {
        val stack = Stack()
        val texture = resourceManager.getTextureAsset(SPRITE_BACKGROUND)
        val background = Image(texture)
        background.name = "background"
        stack.add(background)
        stack.add(createTierColor(NOTHING))
        return stack
    }

    fun select() {
        val texture = resourceManager.getTextureAsset(SPRITE_SELECTED)
        val selector = Image(texture)
        selector.name = "selector"
        super.add(selector)
        tooltip.refresh(this)
    }

    fun deselect() {
        if (isSelected()) {
            super.getChildren().pop()
        } else {
            throw GdxRuntimeException("Tried to deselect an unselected ItemSlot.")
        }
    }

    fun isSelected(): Boolean {
        val name: String? = super.getChildren().peek().name
        return name != null && name == "selector"
    }

    fun getPossibleInventoryImage(): InventoryImage? =
        if (hasItem()) getCertainInventoryImage() else null

    fun getCertainInventoryImage(): InventoryImage =
        super.getChild(1) as InventoryImage

    abstract fun addToStack(actor: Actor)
    abstract fun hasItem(): Boolean
    abstract fun isOnHero(): Boolean
    abstract fun doesAcceptItem(draggedItem: InventoryImage): Boolean
    abstract fun clearStack()
    abstract fun putInSlot(draggedItem: InventoryImage)
    abstract fun getAmount(): Int
    abstract fun incrementAmountBy(amount: Int)
    abstract fun decrementAmountBy(amount: Int)

    fun putItemBack(draggedItem: InventoryImage) {
        putInSlot(draggedItem)
    }

    fun getHalfOfAmount(): Int =
        floor(getAmount() / 2f).toInt()

    fun setItemColor() {
        imagesBackground.children.pop()
        if (hasItem()) {
            imagesBackground.add(createTierColor())
        } else {
            imagesBackground.add(createTierColor(NOTHING))
        }
    }

    private fun createTierColor(): Image {
        val itemId = getCertainInventoryImage().inventoryItem.id
        return when {
            itemId.contains(BASIC) -> createTierColor(BASIC)
            itemId.contains(FINE) -> createTierColor(FINE)
            itemId.contains(SPECIALIST) -> createTierColor(SPECIALIST)
            itemId.contains(MASTERWORK) -> createTierColor(MASTERWORK)
            itemId.contains(EPIC) -> createTierColor(EPIC)
            itemId.contains(LEGENDARY) -> createTierColor(LEGENDARY)
            else -> createTierColor(NOTHING)
        }
    }

    private fun createTierColor(tierColor: String): Image {
        val textureRegion = resourceManager.getAtlasTexture(tierColor)
        val image = Image(textureRegion)
        image.setScaling(Scaling.none)
        image.setPosition(1f, 1f)
        image.name = "tier color"
        return image
    }

}
